### ISSUE TITLE
[API Proposal]: Marshalling strings to unmanaged memory should provide a buffer length

### ISSUE THREAD

**Issue by Mooshua** [dotnet/runtime#81193]
### Background and motivation

Not all strings will follow native norms such as not containing null bytes or properly terminating themselves.

We already handle most of these situations with Native -> CLR Marshal methods, but we don't handle as many going from CLR -> Native.

I am proposing that `ref`/`out` parameters are added to marshal string methods that declare the exact byte length of the unmanaged copy of the string.

In fact, it's explicit behavior declared in `StringToHGlobal____` to copy null bytes embedded in the string. In some scenarios this has no impact (we can calculate the amount of memory copied fairly easily anyways), but we should still provide APIs that minimize the chances for subtle bugs (especially subtle buffer overrun attacks) to appear in what should otherwise be safe unmanaged code.

On top of reducing bugs, this would enable developers to confidently use length-based API functions provided by some unmanaged wrappers to avoid `strlen` calls and drastically increase the performance of unmanaged API invocations--especially in string-laden scenarios.

### API Proposal

```csharp
namespace System.Runtime.InteropServices.Marshal

public class Marshal
{
    // Note: existing string marshal methods use "s" instead of "buffer" here.
    public IntPtr StringToHGlobalAnsi(string? buffer, out int length);
    public IntPtr StringToHGlobalAuto(string? buffer, out int length);
    public IntPtr StringToHGlobalUni(string? buffer, out int length);
    // Why this method is private, I do not know. But it shouldn't be excluded.
    private IntPtr StringToHGlobalUTF8(string? buffer, out int length);

    public IntPtr StringToBSTR(string? buffer, out int length);

    public IntPtr StringToCoTaskMemAnsi(string? buffer, out int length);
    public IntPtr StringToCoTaskMemAuto(string? buffer, out int length);
    public IntPtr StringToCoTaskMemUni(string? buffer, out int length);
    public IntPtr StringToCoTaskMemUTF8(string? buffer, out int length);
}
```

### API Usage

```csharp
//  ================================================================
//  Scenario: Sending string to API that requires length information
//  fetch string
string myString = GetMyString();

//  allocate native copy
var pointer = Marshal.StringToHGlobalAuto(myString, out var length);
var result = NativeAPI.DoWork(pointer, length);

//  ===============================================
//  Scenario: Sending string to API-managed memory.
//  We want to copy the entire buffer using NativeMemory, but NativeMemory and our NativeAPI requires a length parameter.
//  Yes, we can do this by calculating the UTF-16 length, but this introduces less potential for buggy (or insecure) implementations.
string hostileString = GetStringWithNullByte();

//  prepare native version
var unmanagedString = Marshal.StringToHGlobalAuto(hostileString, out var length);
//  the native API now uses our length instead of strlen to avoid falling into the null-byte trap.
var internalString = NativeAPI.AllocateInternalString(length);
//  copy to the API's memory
NativeMemory.Copy(unmanagedString, internalString, length);
//  do work with internal memory
NativeAPI.DoWorkWithInternalString(internalString);
```

### Alternative Designs

Alternatively, APIs that calculate a maximum bound for a native string's length could be provided.

For example, this snippet from `StringToHGlobalAnsi` could be provided as a method `StringToAnsiLength`:
```cs
long lnb = (s.Length + 1) * (long)SystemMaxDBCSCharSize;
int nb = (int)lnb;

// Overflow checking
if (nb != lnb)
{
    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.s);
}
```

### Risks

I do not believe any performance regressions should be caused by these changes. This should break binary compatibility, so new methods should be created instead of modifying the existing methods to keep compatibility.

Additionally, I understand how `out` parameters are heavily discouraged in modern C# APIs. However, I believe the compactness of the syntax (eg, `out var bufflen`), minimal interference, and already unsafe context make it ideal over other solutions.

For example, the changes required for `StringToHGlobalAnsi`:

```cs
//public static unsafe IntPtr StringToHGlobalAnsi(string? s)
public static unsafe IntPtr StringToHGlobalAnsi(string? buffer, out int length)
{
    // ...
    long lnb = (buffer.Length + 1) * (long)SystemMaxDBCSCharSize;
    //int nb = (int)lnb;
    length = (int)lnb;
    // ... Refactor rest of file, replacing "nb" with "length"
}
```

**Comment by Mooshua** [comment]
Hi @AaronRobinsonMSFT! Thank you for the quick response.

At the moment, do you fully approve of the API changes above as-is? If so I believe I can contribute a pull request with these changes.

Thanks.

**Comment by AaronRobinsonMSFT** [comment]
> At the moment, do you fully approve of the API changes above as-is? If so I believe I can contribute a pull request with these changes.

@Mooshua Nope, that isn't my responsibility. I just moved it from "untriaged" to a milestone. In this case, there is no need for the APIs so I simply moved it to Future. I am one of the owners of the interop area and the API seems reasonable. I will say that some of these APIs are very old and newer code should be using tools like [`LibraryImport`](https://learn.microsoft.com/dotnet/standard/native-interop/pinvoke-source-generation) so updating them is less of priority for us. 

> If so I believe I can contribute a pull request with these changes.

After we get consensus, the interop team and community, we will add the "api-ready-for-review" tag and go through an official API review. If the APIs are then approved work can begin. We would welcome the contribution if these are approved though.

Next steps are understanding the full extent of your scenario so we can see how these APIs are needed. For example, why is manually marshalling needed?

**Comment by Mooshua** [comment]
Ah, I see.

A more specific scenario where this change would be useful is binding code to other languages/runtimes:
The motivating example on my end was this export in the Lua API:
```c
//  Internalizes buffer "s" with length "l" and places it on the lua stack.
LUA_API void lua_pushlstring(lua_State* L, const char* s, size_t l);
```
Lua strings can contain *any* data, including null bytes, as they are terminated by a set length, not by a null byte. For this reason, there is no `strlen` version of this function (only a C macro, which is heavily discouraged against using).

**Why not just use a `byte[]`?** 
Because in that case I would be performing an unnecessary copy on arbitrarily sized data. I would much rather move it to and from Lua without having to use an intermediary array copy.

The best thing I can do to calculate the length now is to study the behavior of the specific `HGlobal` string helper function I use and copy it, which does not protect me against breaking changes to this behavior.

The reverse method is already fully supported by the Marshal class:
```c
LUA_API const char* lua_tolstring(lua_State* L, int idx, size_t* len);
```
```cs
// Under the hood, creates ReadOnlySpan<byte>(result, length);
// and feeds that into Encoding.GetString()
value = Marshal.PtrToStringAnsi(result, (int) length);
```

**Comment by AaronRobinsonMSFT** [comment]
> A more specific scenario where this change would be useful is binding code to other languages/runtimes:
> The motivating example on my end was this export in the Lua API:

Ah. This is important to us. Thank you for such a concrete example.

In order to ensure I understand this scenario a current solution here would be to perform something akin to:

```csharp
string dataToSend = "the quick brown fox jumped over the lazy dog";
var buffer = new Span<byte>(NativeMemory.Alloc(1024), 1024);
int encodedLen = Encoding.UTF8.GetBytes(dataToSend.AsSpan(), buffer);
```

The `encodedLen` represents the length of the data that was written into `buffer` and represents a single allocation for the encoding. I _think_ the above is likely what I would recommend since the `Marshal` APIs above dictate the allocator, whereas the above permits the marshalling implementer to decide. This seems far more useful for arbitrary binding code for other languages/runtimes.

**Comment by Mooshua** [comment]
That is a neat trick with `GetBytes` that I didn’t know existed—thank you! 

This solves my motivating scenarios for this API request. Nonetheless I encourage that you still consider it.

Thanks 😄
